### PR TITLE
Fixes berries missing timing

### LINKED ISSUE #5152: Salac Berry Activation Timing issues
### Description

https://github.com/user-attachments/assets/8cfbb9c1-ffab-4eb1-be6d-a545b7569f94

Passive Damage (Toxic Poison in this case) seems to mess with the Activation timing. When Arcanine drops to 16 HP, Salac Berry is supposed to immediately activate. But it only gets activated after using the next move.
Didn't test other berries but more berries are very likely affected.

### Version

1.9.0 (Latest release)

### Upcoming/master Version

_No response_

### Discord contact info

._.terra._.

### PR DESCRIPTION
**Pull request by AlexOn1ine**
Fixes #5152

### CODE REVIEW

**Comment by AsparagusEduardo** [comment]
Can we get a test alongside this fix?

**Comment by AlexOn1ine** [comment]
> Can we get a test alongside this fix?

Ye but it is very likely I will forget so might need a reminder.

**Comment by AsparagusEduardo** [comment]
> Ye but it is very likely I will forget so might need a reminder.

Is it ok that I mark this as a draft then?

**Comment by AlexOn1ine** [comment]
> > Ye but it is very likely I will forget so might need a reminder.
> 
> Is it ok that I mark this as a draft then?
yes

**Comment by AlexOn1ine** [comment]
Added test that fails on master/upcoming + additional missing check.

Sorry for the force push. For reference here is the old commit:
image

File: src/battle_util.c [hunk 1]
```diff
@@ -2453,6 +2454,15 @@ u8 DoBattlerEndTurnEffects(void)
                 effect++;
             gBattleStruct->turnEffectsTracker++;
             break;
+        case ENDTURN_ITEMS3:  // berry effects
+            if (gItemsInfo[gBattleMons[battler].item].pocket == POCKET_BERRIES
+             || GetBattlerHoldEffect(battler, TRUE) == HOLD_EFFECT_RESTORE_HP)
```
Comment (AsparagusEduardo) [inline src/battle_util.c:2459]:
```suggestion
             || GetBattlerHoldEffect(battler, TRUE) == HOLD_EFFECT_RESTORE_HP) // Edge case for Berry Juice
```

File: test/battle/hold_effect/restore_hp.c
[the remark this answers is not in the document]
Comment (AlexOn1ine) [reply test/battle/hold_effect/restore_hp.c:1]: added

**Review by AsparagusEduardo — APPROVED** [review]
Thank you :)
